RFC: SparseVectors, Take 2

mbauman opened the pull request:
This introduces the `SparseVector` to the standard library.  It is the descendent in spirit of #11424, but the entirety of the code is imported from JuliaSparse/SparseVectors.jl, with credit to @lindahua as the primary author.  I've simply been working on getting it to mesh with Base idioms and names.

TODO:
- [x] Get tests passing
- [x] Implement `transpose(::SparseVector)`.  Probably has to just return a dense array for now.
- [x] Rename the module; `module SparseMatrix` no longer makes sense.
- [x] Documentation review: look through the manual for instances where `matrix` needs to be changed to `vector or matrix`.
- [ ] ~~Determine how or if we should deprecate `sparsevec`.  It's a conglomeration of two orthogonal behaviors that already have independent names: `sparse` and `vec`.  Unfortunately, the `sparse` function has so many signatures defined for it there's no unambiguous way to tack on the `sparsevec` functionality for a few of the methods.  Feedback about desired APIs here would be useful.  See [this comment](https://github.com/JuliaLang/julia/pull/13440#issuecomment-145669851) for specifics.~~ **Decision: punt**. The current API works fine for now. We can revisit this later.
- [x] Missing sparse functionality for SparseVector that makes sense:
  - [ ] ~~Define `rowvals(::SparseVector)`? Effectively the same functionality is defined as the currently-non-exported `nonzeroinds`.~~ **Decision: wait** to see if this is needed for writing generic code.
  - [x] `spones(::SparseVector)`
  - [x] `sprandbool(::Integer, ::AbstractFloat)`
- [x] Double-check coverage for large missing segments.  SparseVectors.jl is nicely covered, so I don't anticipate big missing areas here. Report as of 0d9f1a1: https://codecov.io/github/mbauman/julia?ref=a389967eb3fec6742d3037b386bedbd6df0f7d27 (sparse vectors is over 90%). Only large missing block was indexing by a UnitRange; this is now tested.
- [x] Squash all my intermediate commits

lindahua committed:
Import code from JuliaSparse/SparseVectors.jl

mbauman committed:
Integrate SparseVectors.jl with Base sparse module

Renamings:
* Rename sparse module to SparseArrays, and improve its separation from base. This makes it very simple to dynamically reload the sparse module. Move docstrings to their proper place
* _copy_convert → collect
* Rename sparsevector to the existing spzeros and sparsevec.
* Use call overloading instead of call

Remove functionality from SparseVectors.jl:
* Simplify and remove some functors
* Remove SparseVectorView
* Remove no-longer-needed ambiguity preventers

Add functionality for SparseVectors:
* Add similar for SparseVector
* Allow sparsevec(::AbstractArray), not just vectors
* Add spzeros(n); adapt some tests to SparseVector
* Allow CHOLMOD linalg with vectors
* Implement (c)transpose(::SparseVector). Returns a dense vector since a one-row CSC structure is effectively dense but with worse performance.
* Add vector sprandbool and allow passing RNG to all vector sprand* functions. Harden tests against random failures.
* Implement, test and doc spones(::SparseVector)

Improve performance for SparseVector indexing:
* Punt to SparseMatrix for some indexing behaviors. Since the datastructures are compatible and SparseMatrix's routines are more optimized, it is easiest to just construct a temporary SparseMatrix and index into that. This is faster in all but the smallest of cases (N<10)
* Use searchsorted for indexing SparseVector by UnitRange. This is about 20% slower on very small vectors, but is faster in general.

Change SparseMatrix behaviors to take advantage of vectors
* Add indexing behaviors for SparseMatrix->SparseVector
* `vec` and `sparsevec` for CSC return SparseVectors
* Update documentation to incorporate vectors

Minor bugfixes and changes to SparseVectors:
* Compare to literal 0 in vector construction and setindex. This matches SparseMatrix semantics, and makes indexing semantics consistent with regard to stored zeros
* Use checkbounds more thoroughly
* Use error types that are consistent with SparseMatrixCSC
* Minor sparse vector display tweaks. Turn on output limiting by default, remove unused variable `k`, implement and use Base.summary
* Fix missing return, add test

Add some tests:
* Add a test and comment to ensure nobody else tries to share data between vectors and matrices
* Ensure indexing is consistent between one-column sparse matrices and sparse vectors, with special attention to stored zeros.

lindahua commented:
Thanks a lot for working on this!

tkelman commented on base/sparse/sparsevector.jl:
```diff
@@ -0,0 +1,1423 @@
+    end
+    half_screen_rows = limit ? div(rows - 8, 2) : typemax(Int)
+    pad = ndigits(n)
+    k = 0
```
this, and the `k += 1` below are redundant since `k` is the loop variable

mlubin commented on base/sparse/sparsevector.jl:
```diff
@@ -0,0 +1,1423 @@
+    sparsevec(A)
+
+Convert a vector `A` into a sparse vector of size `m`. In julia,
+sparse vectors are really just sparse matrices with one column.
```
Description here should be updated

ViralBShah commented on base/sparse.jl:
```diff
@@ -1,25 +1,27 @@
 importall Base
-importall ..Base.Operators
+importall Base.Operators
 importall Base.LinAlg
```
Everyone knows this, but eventually, we should not have the `importall`.

ViralBShah commented:
How about deprecating `sparsevec`? The `SparseVector` constructor should be good enough to replace it, right?

ScottPJones commented:
You checked off implementing transpose - does that just return a dense array then?
To do it efficiently, would it require implementing support for a SparseCSR structures, as I believe @tbreloff has discussed? (BTW, this is very nice to get better vector support in)

lindahua commented:
Generally, I think ``SparseVector`` may be just a field setter, while functions like ``sparsevec`` or ``sparse`` can do complicated things (such as sorting/combining elements).

I think we probably don't need ``sparsevec``, but let ``sparse`` returns an instance of ``SparseVector`` when there is only one dimension.

ScottPJones commented:
@lindahua That sounds good.  I've been writing a replacement for sparse myself (I am afraid that the one in base might be encumbered by GPL, as it is derived from C source code that was LGPL 2.1, IANAL, so I'm not sure how that plays out legally).  I had had some thoughts about doing transpose of a sparse matrix efficiently, but that isn't something I've implemented in the past.

ViralBShah commented:
I deleted comments that were completely off-topic here.

kshyatt commented on base/sparse/cholmod.jl:
```diff
@@ -1216,6 +1219,17 @@ function cholfact(A::Sparse; kws...)
+doc"""
+    ldltfact(::Union{SparseMatrixCSC,Symmetric{Float64,SparseMatrixCSC{Flaot64,SuiteSparse_long}},Hermitian{Complex{Float64},SparseMatrixCSC{Complex{Float64},SuiteSparse_long}}}; shift=0, perm=Int[]) -> CHOLMOD.Factor
+
+Compute the `LDLt` factorization of a sparse symmetric or Hermitian matrix. A fill-reducing permutation is used. `F = ldltfact(A)` is most frequently used to solve systems of equations `A*x = b` with `F\b`, but also the methods `diag`, `det`, `logdet` are defined for `F`. You can also extract individual factors from `F`, using `F[:L]`. However, since pivoting is on by default, the factorization i ...
```
Wording here is a little awkward - comma splice. Maybe `F = ldltfact(A)` is most frequently used to solve systems of equations `A*x = b` with `F\b`. The factorization also the methods `diag`, `det`, `logdet` defined.`?

Would be nice to have an explanation of what all these factor are (what is `DUP`?)

mbauman replied:
I agree, but I'm not the one to write that documentation.  I just moved the docs here from helpdb.jl.

kshyatt commented on base/sparse/csparse.jl:
```diff
@@ -313,8 +313,17 @@ function csc_permute{Tv,Ti}(A::SparseMatrixCSC{Tv,Ti}, pinv::Vector{Ti}, q::Vect
+    symperm(A, p)
+
+Return the symmetric permutation of `A`, which is `A[p,p]`. `A` should be
+symmetric and sparse, where only the upper triangular part of the matrix is
```
More awkward wording? I'd say "`A` should be symmetric, sparse, and with the upper triangular part stored. Only the upper triangular part of `A` is returned."

kshyatt commented on base/sparse/sparsematrix.jl:
```diff
@@ -44,6 +44,16 @@ modifications to the returned vector will mutate `A` as well. See `rowvals(A)`
+"""
+    rowvals(A)
+
+Return a vector of the row indices of `A`, and any modifications to the returned
```
Comma splice again.

kshyatt commented on base/sparse/sparsematrix.jl:
```diff
@@ -44,6 +44,16 @@ modifications to the returned vector will mutate `A` as well. See `rowvals(A)`
+Return a vector of the row indices of `A`, and any modifications to the returned
+vector will mutate `A` as well. Given the internal storage format of sparse
+matrices, providing access to how the row indices are stored internally can be
+useful in conjuction with iterating over structural nonzero values. See
```
This is a long sentence and also kind of awkward. Not sure how to improve? I might remove everything before the first comma in the "Given..." sentence.

KristofferC commented on base/sparse/sparsematrix.jl:
```diff
@@ -221,6 +231,14 @@ end
 convert(::Type{Matrix}, S::SparseMatrixCSC) = full(S)
 
+
+doc"""
```
Could remove the `doc`, right?

mbauman commented:
> How about deprecating sparsevec? The SparseVector constructor should be good enough to replace it, right?

I think for now it can be split into a combination of the orthogonal names `sparse` and `vec`.  I'm not entirely sure what to do with the `SparseVector` name; it'd be nice to have the constructor mirror `Vector`, but that's awkward until the deprecation for the `SparseMatrix` module goes through.

KristofferC commented on base/sparse/sparsevector.jl:
```diff
@@ -0,0 +1,1441 @@
+            r += 1
+            i2 = I[r]
+            if i2 == i  # accumulate r-th to the l-th entry
+                V[l] = call(combine, V[l], V[r])
```
`combine(V[l], V[r])` to be similar with the `SparseMatrix` construct?

mbauman replied:
Ah, yup, thanks, this comes from SparseVectors.jl's 0.3 support.

mbauman commented:
> I think for now [`sparsevec`] can be split into a combination of the orthogonal names `sparse` and `vec`

So close.  There'd be a clash of meanings with `sparse(::AbstractVector, ::AbstractVector, ::Integer)`.

Is `sparse([1,2],[1,2],3)` a 2x2 matrix with two threes (`[3 0; 0 3]`) or is it a 3-element sparse vector (`[1,2,0]`)?  Any thoughts here?

ScottPJones commented on base/sparse/sparsevector.jl:
```diff
@@ -0,0 +1,1441 @@
+### common.jl
```
Doesn't this need the standard license header? Also the file is not common.jl.

ScottPJones commented on base/sparse/sparsevector.jl:
```diff
@@ -0,0 +1,1441 @@
+                break
+            end
+        end
+        r > m && SparseVector(len, Ti[], Tv[])
```
This line doesn't make sense to me, was it supposed to be:
`r > m && return SparseVector(...)`?

ScottPJones replied:
@mbauman Did I miss something, it seems that that statement must be incorrect?
It creates a SparseVector and just throws it away.

mbauman replied:
Please don't ping twice within an hour.  I was working my way through all review comments.  I would have much preferred the second message to have real content (like a test case).

KristofferC commented on base/sparse/sparsevector.jl:
```diff
@@ -0,0 +1,1441 @@
+        v = s[i]
+        if v != 0
+            if c >= cap
+                cap *= 2
```
I benchmarked this a little bit and the manual resizing seems to be consistently a bit faster than just pushing. I am a bit surprised though, isn't this exactly what `push!` should do "behind the scenes"?

mbauman replied:
Did you try using `sizehint!(nzind, cap)`, etc?  I agree it'd be better to use push here, but thanks for benchmarking it!

KristofferC replied:
I just used tried with the most simple: https://gist.github.com/KristofferC/2e511458a213aa8c6d85

mbauman replied:
I don't think that's quite correct.  You want to start with zero-length vectors.  You're pushing on top of uninitialized memory.

KristofferC replied:
Oh yes of course. Sorry about that hehe.

KristofferC commented on base/sparse/sparsevector.jl:
```diff
@@ -0,0 +1,1441 @@
+        ix = xnzind[xj]
+        iy = ynzind[yj]
+        if ix == iy
+            s += call(f, xnzval[xj], ynzval[yj])
```
`call` here to

mbauman replied:
Yeah, there's lots.  I've gotten them all, I think.  I'll push.

ScottPJones commented:
I'd asked before about transposing a sparse vector (but Viral deleted question), why couldn't that be returned as a sparse matrix?  It wouldn't be quite as compact, but I'd think it would still be smaller than a dense array, or wouldn't it?

mbauman commented:
Our sparse matrices are CSC format.  I'd be good for you to thoroughly understand how CSC works if you're interesting in contributing to our sparse library.

```jl
julia> S = sparse([1,10^7],[1,1],[1,1])
10000000x1 sparse matrix with 2 Int64 entries:
	[1       ,        1]  =  1
	[10000000,        1]  =  1

julia> Base.summarysize(S)
88

julia> Base.summarysize(S')
80000080

julia> Base.summarysize(zeros(10^7))
80000000
```

ScottPJones commented:
I do understand perfectly well how the CSC format works.
Try the following:
S = sparse(Int32[1,10^7],Int32[1,1],[10,20])
Base.summarysize(S)
You will then see that in that case, you can save almost 50% of the memory.
However, this is really dependent on the # of zero entries, and the type of the value array.
S = sparse(Int32[1,10^7],Int32[1,1],Int128[10,20])
Base.summarysize(S)
Compare that to Base.summarysize(zeros(Int128,10^7)), and you'll see the difference.

ScottPJones commented on test/sparsedir/sparsevector.jl:
```diff
@@ -0,0 +1,649 @@
+## sparsevec.jl
```
This also has incorrect name and is missing Julia license info.

lindahua commented:
Looks really great!

jakebolewski commented:
I see that you replaced the comparison `Ty = eltype(SparseVector(...)); v != zero(Ty)` with `v != 0`.  Does this assumption (not assuming a `zero` method for a given element type) actually hold throughout the sparsematrix code?  I feel that comparison to numeric zero has been gradually removed as it can cause perf problems due to type instability.

mbauman commented:
The SparseMatrixCSC code is a little mixed as to whether it compares to `zero(T)` or `0`.  These changes were required to get tests to pass. At least some of those tests fail because they store `Char`, and probably date back to the days when `Char <: Integer`.

jakebolewski commented:
I think we should be more explicit as to what the interface is assumed rather than having partially implemented ill defined behavior.

ex:
```julia
julia> A = sparse([1,5,10],[1,5,10],['a','b','c'])
10x10 sparse matrix with 3 Char entries:
	[1 ,  1]  =  'a'
	[5 ,  5]  =  'b'
	[10, 10]  =  'c'

julia> full(A)
ERROR: MethodError: `zero` has no method matching zero(::Type{Char})
 in full at sparse/sparsematrix.jl:227


julia> A[1]
'a'

julia> A[2]
ERROR: MethodError: `zero` has no method matching zero(::Type{Char})
 in getindex at sparse/sparsematrix.jl:1398
 in getindex at abstractarray.jl:483
```

mbauman commented:
So the tests that required comparisons to literal zeros were all sparse matrices of `Char`s.  We could either remove those tests, or we could define `zero(::Char)`.  I went the latter direction, since addition is also defined on `Char` and so there's some sense to also supporting the additive identity.

I don't want to make broader changes to SparseMatrixCSC in the PR - I think that's out of scope for this issue.

ScottPJones commented:
instead of defining zero(::Char), which doesn't really make much sense, logically, what about adding a
`is_nonsparse(val)`, which by default would fallback to:
`is_nonsparse(val::T) = val != zero(T)`
That would allow people to define `is_nonsparse` methods that make more sense for their types.
For example, you could define a `is_nonsparse` for strings that considers the zero-length string as being "sparse" for this function, or one for arrays, etc.

mbauman commented:
As we've already stated, larger redesigns are not in scope for this issue.  The goal here is simply to get SparseVector to mesh well with existing functionality.

IainNZ commented:
On the topic of 0 vs `zero`, I tried to make it more consistent once but we got stuck in a quagmire: https://github.com/JuliaLang/julia/pull/9325

lindahua commented:
I agree the `zero` issue should be left to other issues.

andreasnoack commented:
@IainNZ Regarding your issue, I think the solution `x==zero(x)` was dismissed too quickly. It might be inefficient because the compiler always gives up on arrays or nonbits number types like `BigFloat`, but I believe it should work generically in contrast to `x==zero(typeof(x))` which fails for arrays.

I cannot see a problem with defining `zero(::Type{Char})==0` when
```
julia> '\0' + 0 === '\0'
true
```
and
```
help?> zero
search: zero zeros spzeros nonzeros set_zero_subnormals get_zero_subnormals count_zeros RoundToZero leading_zeros

  zero(x)

  Get the additive identity element for the type of x (x can also specify the type itself).
```

StefanKarpinski commented:
Defining zero(Char) isn't quite kosher since we don't define addition of characters but I guess it's fairly harmless. Why do we need sparse vectors of Chars? And why would '\0' make sense as a default value in such a vector?

ScottPJones commented:
Also, is it for sure that directly doing additions and subtractions of `Char`s will be allowed in the future?
This is why I think the idea of having a separate function (whatever the name is), that isn't tied to an idea of zero-ness, would be useful, esp. if you could pass it in to a sparse constructor, like you can do with the combine function.  That could be quite useful to make sparse arrays, where the "sparseness" is say, being a `NaN`, `Null`, the zero-length string, or being a whitespace or control character, whatever you want it to be.  I also believe that a `isnonsparse` method could be a lot faster for things like `BigInt`s and `BigFloat`s (or other user defined types) where constructing a zero value and comparing to it might be 1) expensive or 2) depend on a particular meaning for == (for example, do you want -0.0s to be removed or not).

mbauman commented:
Yes, I'm tackling this a different way now which will match the existing SparseMatrix code a little better.  In general, SparseVectors.jl is more aggressive at removing stored zeros than SparseMatrixCSC is.  Making this match better will also remove the need for `zero(::Char)`.

mbauman commented:
Alright, updated.

* I've removed `zero(::Char)`.  While subtraction is defined between two Chars, addition is not.
* I've matched SparseMatrix semantics for `setindex!` and construction with regards to comparisons against literal zero.  This is a compromise — it leaves comparisons against `zero(T)` in other contexts, but ensures that indexing and construction semantics match between matrices and vectors.  A more systematic change can be made later with regards to how sparse arrays handle `zero`.

mbauman committed:
Add SparseVector to NEWS.md

mbauman committed:
Move SparseArrays exports back to their usual home

mbauman commented:
Thank you all for the review comments so far.  Unless anyone has anything else, this is now ready as far as I'm concerned.  I'm sure there will be a long tail of missing methods, performance improvements, and perhaps some bugs, but I think it's really quite good.  It's well-covered by tests, and behaves as I'd expect.

mbauman committed:
Add missing imports

mbauman commented:
Both 64 bit linux workers got hit by the OOM killer.

jakebolewski commented:
:cake:

KristofferC commented:
Great!

malmaud commented:
:1234: 

Is there a constructor to create an all-zero sparse matrix of a given size? My brief foray didn't reveal one.

StefanKarpinski commented:
spzeros?

malmaud commented:
yep, thanks.

On Wed, Oct 14, 2015 at 4:54 AM, Stefan Karpinski <notifications@github.com>
wrote:

> spzeros?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/13440#issuecomment-147980620>.
>

lindahua commented:
Great, thanks a lot!

ViralBShah commented:
This is amazing! Thanks everyone. Kicking the tires.
